### ISSUE TITLE
Activations become NaN?

### ISSUE THREAD

**Issue by juesato** [Element-Research/rnn#187]
I'm really sorry to ask for help about such an underspecified bug (is there a better place to ask for help) - I was wondering if anybody has experience with similar issues.

I'm training an LSTM network (with MaskZero if that's relevant), and after training for a while, the outputs become NaN (and the weights as well). I thought it could be an exploding gradient at first, but I don't think that's the case, since I'm monitoring the max value with the gradParams tensor and nothing weird is happening there.

I also don't think it's an input formatting issue, since the bug appears at different training steps, depending on how I set the random weight initialization, or the learning rate.

Does anybody have suggestions for how to debug this? I'd really appreciate it.

**Comment by nicholas-leonard** [comment]
@juesato NaN bugs are the worst. The way I debug them is I try to narrow down where it gets created. So clone nn and rnn and whatnot. And then I do something like this to the modules that might be responsible. Using `Linear` as an example of a possible culprit : 
```lua
local _ = require 'moses'
local updateOutput = nn.Linear.updateOutput
function nn.Linear.updateOutput(self,  input)
    assert(not _.isNaN(input:sum()))
    local output = updateOutput(self, input)
    assert(not _.isNan(output:sum())
    return output
end
```

Typical causes of NaNs are missing `tensor:zero()` before some operations like indexCopy or addmm, or log(0). If you can narrow down the problem to a module, I might be able to help you. I just reviewed the MaskZero and MaskZeroCriterion code for any obvious NaN bugs and couldn't find any.

**Comment by kmul00** [comment]
Hi,

Just to add to what Nicholas has already said, quite a few times when I have encountered `NaN`, I got rid of it by changing the learning rate (reducing it worked in my case).

**Comment by juesato** [comment]
@nicholas-leonard Thanks for the help. I overwrote the `updateOutput` functions for each of my layers like you suggested, and the AssertionError occurred in `LookupTableMaskZero`.

On one value of the learning rate, I got a "bus error," which I hadn't seen before. I'm not sure if this is relevant, or a separate bug.

@kmul00 I've been varying the learning rate, and the bug ends up showing up at different time steps. I'm also sceptical it's an exploding gradient issue since I'm monitoring the gradients, so I'm not sure lowering the learning rate should help. Does this match up with what your experience?

**Comment by juesato** [comment]
@nicholas-leonard As an update, I believe the NaN occurs in the backwards pass

I used similar code to test the gradients

`local accGradParametersLinear = nn.Linear.accGradParameters

function nn.Linear.accGradParameters(self, input, gradOutput, scale)
    assert(not _.isNaN(gradOutput:sum()))
    local gradInput = accGradParametersLinear(self, input, gradOutput, scale)
-- gradInput is null
    assert(not _.isNaN(self.weight))
    assert(not _.isNaN(self.gradWeight))
end

local updateGradInputSoftmax = nn.LogSoftMax.updateGradInput

function nn.LogSoftMax.updateGradInput(self, input, gradOutput)
    assert(not _.isNaN(gradOutput:sum()))
    local gradInput = updateGradInputSoftmax(self, input, gradOutput)
    assert(not _.isNaN(gradInput:sum()))
    return gradInput
end
`

When I run this, the assertion error occurred in the first line, inside `Linear` on `assert(not _.isNaN(gradOutput:sum()))`. (This is a little confusing to me, since I would expect the `gradInput` from the softmax layer to be the same as the `gradOutput` of the Linear layer, given that my network is LookupTable -> LSTM -> Linear -> Softmax).

**Comment by juesato** [comment]
I've been calling the `AbstractRecurrent:maskZero()` function, which is currently implemented as

```
function AbstractRecurrent:maskZero(nInputDim)
   self.recurrentModule = nn.MaskZero(self.recurrentModule, nInputDim, true)
   self.sharedClones = {self.recurrentModule}
   return self
end
```

I noticed that in `LSTM.lua`, the variable `self.recurrentModule` is also used for `self.modules` (https://github.com/Element-Research/rnn/blob/master/LSTM.lua#L22) so maybe these needs to be overrode for `LSTM` as

```
function AbstractRecurrent:maskZero(nInputDim)
   self.recurrentModule = nn.MaskZero(self.recurrentModule, nInputDim, true)
   self.sharedClones = {self.recurrentModule}
   self.modules = {self.recurrentModule}
   return self
end
```

Do you think this would fix the bug?

**Comment by nicholas-leonard** [comment]
@juesato I modified LookupTableMaskZero so that it uses a buffer instead of allocating memory (via `torch.add`) every call to forward. Not sure that will solve any problems, but it will help control memory allocation. I am now looking into the remainder of your comments.

**Comment by juesato** [comment]
@nicholas-leonard Thanks. I don't think that changed the bug causing the NaN's. 

I don't think my second comment about `AbstractRecurrent:maskZero()` is the source of my bug, though it does seem like `self.modules` should point at the masked version of the module (maybe this doesn't matter?).

I'm not sure what other approaches I can use to debug this.

**Comment by nicholas-leonard** [comment]
@juesato Another good point. I fixed it in https://github.com/Element-Research/rnn/commit/091d24cfa4a728ce92c6393914c9c805417297fd .

In any case, if Linear's gradOutput contains a NaN, it means that another module is the cause of it. One of the modules after a Linear in the LSTM maybe.

**Comment by nicholas-leonard** [comment]
@juesato Also, try out the new [`nn.NaN`](https://github.com/Element-Research/dpnn#nn.NaN) from dpnn.

**Comment by juesato** [comment]
@nicholas-leonard Thanks, I've wrapped all my layers with `nn.NaN`. I'm pretty sure the issue is coming up in the LSTM layer (depending on different things I try, the error message looks different, but it always seems to come back to the LSTM layer.

```
In 3 module of nn.Sequential:
...edu/u/j/juesato/torch/install/share/lua/5.1/dpnn/NaN.lua:52: NaN found in gra
dInput of module :
nn.NaN(3) @ nn.Sequencer @ nn.LSTM(256 -> 256)
```

**Comment by nicholas-leonard** [comment]
@juesato LSTM contains many modules. Try something like this to get additional granularity : https://github.com/nicholas-leonard/rnn/blob/noise-contrastive-estimate/examples/noise-contrastive-estimate.lua#L134-L155 .

**Comment by juesato** [comment]
The results seem inconsistent, but the error is always in the LSTM/GRU layer. Sometimes it's in gradInput, sometimes it's in output.

Here's one trace:

```
$ th train3_uni.lua -ch
eckpoint_dir cv/13apr4_uni_gru/ > cv/13apr4_uni_gru/log.txt
/afs/csail.mit.edu/u/j/juesato/torch/install/bin/luajit: ...u/j/juesato/torch/in
stall/share/lua/5.1/nn/Container.lua:69: 
In 2 module of nn.Sequential:
In 1 module of nn.Sequential:
In 2 module of nn.ConcatTable:
In 1 module of nn.Sequential:
In 1 module of nn.ParallelTable:
```

This points to the `nn.Linear` layer here: https://github.com/Element-Research/rnn/blob/master/GRU.lua#L65 . This doesn't make sense to me though, because I don't understand how none of the parameters have NaN, and none of the input have NaN, but the output has NaN in a linear layer.

**Comment by nicholas-leonard** [comment]
Wow that is a pickle. [Linear](#https://github.com/torch/nn/blob/master/Linear.lua) looks fine as far as I can tell. Your nn package is up to date right? 

Otherwise, I don't understand how to generate a NaN from an affine transform that contains none. Maybe try modifying the NaN module locally so that it saves the decorated module and its input on disk when the error occurs. That way you could inspect the contents.

**Comment by juesato** [comment]
@nicholas-leonard My mistake - it was in fact an exploding gradient. I was only checking the max value of the gradients every 10 iterations, and it was exploding really quickly, going from ~10 to 1e20 in 4 or 5 iterations, so I wasn't catching that. Sorry for the trouble.

**Comment by nicholas-leonard** [comment]
@juesato Glad you could identify the issue! Those pesky NaN bugs...

**Comment by NightFury13** [comment]
@juesato : so what finally solved your problem? I am stuck with a very similar situation. I am trying to do a sequence learning task with a BiLSTM-RNN+CTC-layer. However, the activations and ctc-error become NaN after the very first iteration.

```
The RNN I'm using :

nn.Sequential {
  [input -> (1) -> (2) -> (3) -> output]
  (1): nn.Sequential {
    [input -> (1) -> (2) -> (3) -> output]
    (1): nn.ParallelTable {
      input
        |`-> (1): nn.Sequencer @ nn.Recursor @ nn.Sequential {
        |      [input -> (1) -> (2) -> output]
        |      (1): nn.LookupTableMaskZero
        |      (2): nn.FastLSTM(50 -> 50)
        |    }
        |`-> (2): nn.Sequencer @ nn.Recursor @ nn.Sequential {
        |      [input -> (1) -> (2) -> output]
        |      (1): nn.LookupTableMaskZero
        |      (2): nn.FastLSTM(50 -> 50)
        |    }
         ... -> output
    }
    (2): nn.ZipTable
    (3): nn.Sequencer @ nn.Recursor @ nn.JoinTable
  }
  (2): nn.Sequencer @ nn.Recursor @ nn.MaskZero @ nn.Linear(100 -> 109)
  (3): nn.Sequencer @ nn.Recursor @ nn.MaskZero @ nn.LogSoftMax
}
```

**Comment by juesato** [comment]
@NightFury13 It sounds like you have a different issue if you're getting issues on the first iteration, but I'll try my best to help out. My thinking is that if the NaN is on the first iteration there's probably just a syntax error in the way the network is set up.

Have you set up the input and output dimensions on your LookupTableMaskZero? (I only ask because it's not included in the output). What happens if you use an ordinary LookupTable instead?

**Comment by NightFury13** [comment]
@juesato : thanks for responding! :D 

I get the NaN not from the first, but from the second iteration onwards. That change things a bit? 

I set up the input/output dimensions in the LookupTableMaskZero, "local sharedLookupTable = nn.LookupTableMaskZero(nIndex, hiddenSize)" & "add(nn.Sequencer(nn.MaskZero(nn.Linear(hiddenSize*2, nVocab), 1)))". Below is my model definition, is there anything off with this?

```
function models.monoBatch_BiLSTM(nIndex, hiddensize, nVocab)
    local sharedLookupTable = nn.LookupTableMaskZero(nIndex, hiddenSize)

    -- forward rnn
    local fwd = nn.Sequential()
        :add(sharedLookupTable)
        :add(nn.FastLSTM(hiddenSize, hiddenSize):maskZero(1))

    -- internally, rnn will be wrapped into a Recursor to make it an AbstractRecurrent instance.
    fwdSeq = nn.Sequencer(fwd)

    -- backward rnn (will be applied in reverse order of input sequence)
    local bwd = nn.Sequential()
        :add(sharedLookupTable:sharedClone())
        :add(nn.FastLSTM(hiddenSize, hiddenSize):maskZero(1))

    bwdSeq = nn.Sequencer(bwd)

    -- merges the output of one time-step of fwd and bwd rnns.
    -- You could also try nn.AddTable(), nn.Identity(), etc.
    local merge = nn.JoinTable(1, 1)
    mergeSeq = nn.Sequencer(merge)

    -- Assume that two input sequences are given (original and reverse, both are right-padded).
    -- Instead of ConcatTable, we use ParallelTable here.
    local parallel = nn.ParallelTable()
    parallel:add(fwdSeq):add(bwdSeq)
    local brnn = nn.Sequential()
        :add(parallel)
        :add(nn.ZipTable())
        :add(mergeSeq)

    local rnn = nn.Sequential()
        :add(brnn)
        :add(nn.Sequencer(nn.MaskZero(nn.Linear(hiddenSize*2, nVocab), 1))) -- times two due to JoinTable
        :add(nn.Sequencer(nn.MaskZero(nn.LogSoftMax(), 1))) 

    return rnn
end
```

**Comment by NightFury13** [comment]
Found the problem atleast. My gradients are exploding!
```
[INFO] : Model Successfully Initialized.	
[INFO] : Training the model.	
[0.001 epoch] : Iteration 1; Error 62.110218	
0 0 0	------ This is before the rnn:backward() call. [expected due to rnn:zeroGradParameters()]
inf -1.7695874380504e+307 5.0816263541685e+306  ---- This is after rnn:backward() call. :"(
[0.002 epoch] : Iteration 2; Error nan	
0 0 0	
nan nan nan	
[0.003 epoch] : Iteration 3; Error nan
```
I am only re-creating a problem solved by someone else using alex-graves' C++ libraries so the data I have is correct. This means the model is faulty?

What I need is a Bi-directional LSTM based RNN, and this model I took from the examples of the RNN-module for torch itself. [[LINK](https://github.com/Element-Research/rnn/blob/master/examples/simple-bisequencer-network-variable.lua)]. Would there be better ways of defining the model?

**Comment by JoostvDoorn** [comment]
Why don't you use BiSequencer?

**Comment by NightFury13** [comment]
There was no specific reason as such. I felt both of them were doing pretty much (if not exactly) the same thing. I however felt, that the reverse-input-sequence part was more trivial in the previous case (refer above comments for code) than the BiSequencer. Will it have a difference? Apologies for the lack of knowledge in this field.

PS : This is what you mean by a BiSequencer right?
```
-- forward rnn
-- build simple recurrent neural network
local fwd = nn.Recurrent(
   hiddenSize, nn.LookupTable(nIndex, hiddenSize), 
   nn.Linear(hiddenSize, hiddenSize), nn.Sigmoid(), 
   rho
)

-- backward rnn (will be applied in reverse order of input sequence)
local bwd = fwd:clone()
bwd:reset() -- reinitializes parameters

-- merges the output of one time-step of fwd and bwd rnns.
-- You could also try nn.AddTable(), nn.Identity(), etc.
local merge = nn.JoinTable(1, 1) 

-- we use BiSequencerLM because this is a language model (previous and next words to predict current word).
-- If we used BiSequencer, x[t] would be used to predict y[t] = x[t] (which is cheating).
-- Note that bwd and merge argument are optional and will default to the above.
local brnn = nn.BiSequencerLM(fwd, bwd, merge)

local rnn = nn.Sequential()
   :add(brnn) 
   :add(nn.Sequencer(nn.Linear(hiddenSize*2, nIndex))) -- times two due to JoinTable
   :add(nn.Sequencer(nn.LogSoftMax()))
```

**Comment by nicholas-leonard** [comment]
@NightFury13 I am assuming you are calling `zeroGradParameters` before the backward. If you have a really big vocabulary (100k+ words), you might need to initialize the `linear.bias` to something like  `-log(output vocab size)`. Another solution is to do gradient clipping : https://github.com/Element-Research/rnn/blob/master/examples/recurrent-language-model.lua#L206-209. A norm of 5 or 10 usually does the trick.

**Comment by NightFury13** [comment]
@nicholas-leonard : My vocabulary is actually only 109 chars (unitext-codes of english-alphabet and a UNK char). I think the problem was in my model definition. I simply removed the LookupTable()'s and it seems to be working fine now.

```
function models.monoBatch_BiSeqLSTM(nInput, hiddenSize, nOutput)
    local rnn = nn.Sequential()
        :add(nn.BiSequencer(nn.FastLSTM(hiddenSize,hiddenSize)))
        :add(nn.Sequencer(nn.Linear(hiddenSize*2, nOutput)))
        :add(nn.Sequencer(nn.LogSoftMax()))

    return rnn
end
```
Not an expert in torch yet, so I am not sure what problem is getting solved doing so. From what I read online, the LookupTableMaskZero helps when you have inputs of varying sequence lengths and this automatically pads 0's for the extra bits. Can you explain/guess what might be going wrong and exploding the gradients to such huge values?

As for clipping, I don't think that would help my case either as the minimum value itself is near inf. So I can't really cut off values above a threshold, unless I keep the threshold to be inf itself. 
```
[0.001 epoch] : Iteration 1; Error 62.110218    
0 0 0   ------ This is before the rnn:backward() call. [expected due to rnn:zeroGradParameters()]
inf -1.7695874380504e+307 5.0816263541685e+306  ---- This is after rnn:backward() call.
  |                      |                                    |  
  |                      |                                    |
 norm              min                                max
```

**Comment by nicholas-leonard** [comment]
@NightFury13 Ah so you are doing a character-level language model. Because the vocabulary is so small, you can use [`nn.OneHot`](https://github.com/Element-Research/dpnn/#nn.OneHot) instead of `nn.LookupTable`. The mask-zero version is only useful if you separate independent sequences with zero-indices. 

It is hard to say what causes the explosion. I could be the criterion, bad initialization of the parameters, or too long a sequence length. Typically, a log or exp will amplify the values into inf or nans. I am having similar issues for noise-contrastive estimation. Gradient clipping fixes it for me, but I am trying to better initialize the bias instead.

**Comment by NightFury13** [comment]
Thanks for the tip @nicholas-leonard ! Will try out nn.OneHot and update with findings.

I am using the CTC-criterion. Its actually a char-level-OCR-task so the CTC is a good candidate. I have fixed-height and varying width images (rescaled to fixed height and hence varying width) and the model learns to convert them to their corresponding unicode. I feed one image column (fix-height) at each time step, so the seq-length (width of img) varies with each sample.

**Comment by nicholas-leonard** [comment]
@NightFury13 Ok so your input is an an image column and your output is a class (unicode char). In that case, neither LookupTable nor OneHot is appropriate. For you, the input can just be the BiSequencer LSTM which is why the above fixes your NaNs :) Please let me know if you open source your code, will add it to rnn list of external resources.
